### ISSUE TITLE
Elevation correction don't work

### ISSUE THREAD

**Issue by Vucumpra** [osmandapp/OsmAnd#21418]
### Description

Elevation correction don't work since update 

### Steps to reproduce

Record a track.
Open the track and select "elevation correction" on the menu.
Then select "get elevation data" > "Use terrain maps"

### Actual result

The data will be exactly the same

### Expected result

Register,  into gpx file, all new elevation data from terrain maps.

### Your Environment (required)

**WARNING** Crash-Logs **MAY** contain information you deem sensitive.
Review this **CAREFULLY** before posting your issue!

```
OsmAnd Version: OsmAnd+ 4.9.6 F-Droid
Android/iOS version: Android 11
Device model: 1+ Nord 100
Crash-Logs: ? no
```

**Comment by yuriiurshuliak** [comment]
The bug has been reproduced under the following steps:  
1. Record a track in OsmAnd.  
2. Open the recorded track and select **Options > Elevation correction > Graph**.  
3. Note the altitude data displayed on the graph.  
4. Select **Get elevation data > Use Terrain maps**.  
5. Check the graph again.  

**Result**: The altitude data remains unchanged.  

**Observations:**

- In my testing, the altitude did not significantly change on a track recorded in a flat city area. However, on tracks with sharp altitude variations (e.g., in mountainous regions), the feature might yield different results.  
- Interestingly, I also tested this feature on unrecorded tracks (imported GPX files), and in these cases, the altitude data updated as expected.  

OsmAnd~ 5.0.0#3793m, released: 2024-11-25

https://github.com/user-attachments/assets/6628e089-9243-47d5-abf8-d4506ef8a58a

**Comment by nazar-kutz** [comment]
Further investigation of the bug revealed the cause of the point count mismatch between the GPX file and `editingContext`.

The issue occurs because tracks recorded using "Trip Recording" do not include "route" information (stored in the `<rte>` tag in the GPX structure). For such tracks, in the `segmentForSnap` method, all segment points (except the first and last ones) are duplicated:  
[[Code Reference](https://github.com/osmandapp/OsmAnd/blob/0cabf8d2cafd71eb42fd012e7e3c4777f19008b2/OsmAnd/src/net/osmand/plus/measurementtool/MeasurementEditingContext.java#L842)](https://github.com/osmandapp/OsmAnd/blob/0cabf8d2cafd71eb42fd012e7e3c4777f19008b2/OsmAnd/src/net/osmand/plus/measurementtool/MeasurementEditingContext.java#L842)

This duplication might have been implemented for a specific purpose, but it causes issues for the Elevation Correction function.

#### Current Tasks:
1. Investigate the purpose of the point duplication.  
2. If duplication is unnecessary, remove it to prevent bugs in Elevation Correction and other functions.  
3. If duplication is essential for other functions, fix the Elevation Correction bug by implementing an alternative check, such as comparing points based on their coordinates instead of their count before found elevation data will be aplied.

**Comment by Zirochkabila** [comment]
# Testing 1
`OsmAnd~ 5.0.0#4148m, Released: 2025-01-16` - Main issue **Fixed**

When selecting **Get elevation data** > **Use Terrain maps** and returning to the main screen, the data in the track graph is updated (the same as in **Use Terrain maps**) even **without using** Use Terrain maps (by pressing Done) until the application **is restarted**

- [ ] Updated graph **should not be saved** unless it is **Done** is selected in the **Use Terrain maps** menu


<details><summary>Updated graphic</summary>
<video src="https://github.com/user-attachments/assets/d3cf7950-2c96-459a-a5d2-7dd251aafce8"></video>
</details>

**Comment by sonora** [comment]
@vshcherb ,  @nazar-kutz 

Not quite sure if I misunderstand some of what is requested or decided above, but please note that as per GPX GPS Exchange Format the `<ele>` tag in a gpx file represents the elevation of a point, **measured in meters, expressed as a decimal number, and using the WGS 84 datum**.

Should we somehow consider to factor in elevation corrections and write them back to the gpx files' `<ele>` tag, I believe this would be a violation of the GPX format. Such corrections would require being written to custom extensions, if at all...

**Comment by vshcherb** [comment]
I don't have an answer, I feel we could write a tag source of elevation in extensions for the whole gpx and somehow display in UI what's source of Elevation?

**Comment by sonora** [comment]
Yes, I've seen things like this
```
<gpx version="1.1" creator="XYZ">
  <metadata>
    <desc>Elevation data is referenced to the EGM96 geoid model.</desc>
  </metadata>
  <wpt lat="...
  </wpt>
</gpx>
```
but am not so sure it is really in line with GPX 1.1.

**Comment by vshcherb** [comment]
What does it mean geoidheight ? I believe the best should be specified by geoid EGM96 or EGM2020 is coming somewhere in extensions of gpx itself.
WGS 84 - is a system of projection but doesn't specify which geoid is used. 
As we go more professional side we need to understand which geoid GPS chipset operates to compensate to another geoid model (altitude correction file) when we use elevation correction it uses DEM data which I believe doesn't EGM84 but rather different model however in many places its precision of 30-60 m per pixel is just too much

**Comment by sonora** [comment]
My comment was strictly about notation, let me rephrase what I am saying: The `ele` tag in a GPX file per default specification contains elevation data in meters **relative to the WGS84 ellipsoid** (which is the standard coordinate system used for GPS). (In the environment I am working in we very strictly use WGS84 ele tagging.)

If  in exceptional cases the `ele` tag data is referenced to a different datum, this should be clearly specified in the GPX file's metadata. But even then, there it is easily overlooked, and such a GPX file may be misinterpreted by most software.

A less errorprone method may be to ensure the `ele` tag only holds the standard 'altitude vs.WGS84' component of each `wpt`, and any local correction (obtained by using a different datum or from a correction file applied) is reported in a separate correction tag (something like local geoidheight or similar extension tag).

Of course you are right that with the advent of devices reporting inherently corrected (non WGS84) elevation values, we are now challenged to discover this in order to still write unambiguous GPX files. Probaly a small research project to see if this can somehow be done for all the different hardware scenarios. I am a little pessimistic there. Professional devices do of course document this, for consumer devices it could be tricky.

**Comment by vshcherb** [comment]
I mean we were not very consistent in the past, so we need to review the DEM data even precision of the data had much bigger impact. Usually indeed vertical precision is much higher than horizontal in GPS and that accuracy is getting very high with lots of satellites available. 

We need to find a tool that would show difference WGS84 and other datum per location. However I checked few location difference is only 15 cm - https://geographiclib.sourceforge.io/cgi-bin/GeoidEval?input=50.439720+30.514080&option=Submit

**Comment by vshcherb** [comment]
So if I understand the elevation expected over ellipsoid even though this not the elevation many would see as EGM84 has difference of 5 m to the ellipsoid itself.

**Comment by sonora** [comment]
> We need to find a tool that would show difference WGS84 and other datum per location. However I checked few location difference is only 15 cm - https://geographiclib.sourceforge.io/cgi-bin/GeoidEval?input=50.439720+30.514080&option=Submit

From what I recall, our planet's geoid deviation vs. WGS84 is in the range **-106 meters to +85 meters**.

Your linked position produces **+25 m**: (Please note that WGS84, an ellispid model, is different from EGM84, a gravitational 'potato' model :wink:)

![Image](https://github.com/user-attachments/assets/16265c66-75bc-448f-9275-69dce295a106)

**Comment by vshcherb** [comment]
Elevation correction now works though for providing source of elevation we could create a separate issue

**Comment by DavidHenryThoreau** [comment]
I'm using latest F-Droid package : https://f-droid.org/repo/net.osmand.plus_500503.apk 
I'm getting `404 Not Found`

**Comment by DmitryAlexei** [comment]
@DavidHenryThoreau Online calculation is no longer available.
You can download the latest build from Google Play and utilize the Elevation calculation feature offline, as outlined in the documentation: https://osmand.net/docs/user/map/tracks/track-context-menu#calculate-missing-elevation

**Comment by DavidHenryThoreau** [comment]
Thanks for reply.
I've switched to local elevation feature, i've made some test with old gpx which elevation gain is 120 after online correction, after used your documentation https://osmand.net/docs/user/map/tracks/track-context-menu/#calculate-missing-elevation I'm getting 160 elevation gain.
I've also testing with another track the elevation does not move after correction, is that normal ?

**Comment by sonora** [comment]
@DavidHenryThoreau I am not quite sure, but (since you say you used 'old' gpx for this test), probably the elevation correction is only applied if the track file itself does not already contain elevation data (ele tags), which would have been added when the track was e.g. recorded. I think the mechanism you quoted is only intended for when you are engineer a new gpx file from taps on the map, or have an existing one (like from a recording or a guide book) which is missing any kind of elevation information.

**Comment by DavidHenryThoreau** [comment]
@sonora I've recorded new tracks after following the documentation steps, the elevation does not vary after recalculation.
Another precision these tracks are recorded in plane mode (no data) in order to preserve battery percentage.

**Comment by sonora** [comment]
@DavidHenryThoreau Perhaps I had not quite understood your use case and procedure, so probably my comment above is not very helpful or even false for what you're doing.

For all practical matters, if you are
- on the ground (walking, driving, i.e. not in an airplane) and record a track in OsmAnd,
- and are interested in elevation **gain and loss** of that trip, i.e. delta values,

I believe the data as is recorded is typically better (more precise) than any successive matching with any terrain data 'correction'.

Correction probably is only useful
- if you have a track of other origin with e.g. no elevation data contained
- or if you are interested not in gain/loss but in absolutue values and your device's GPS system provides values not corrected to our planet's varying mean sea level. This case can be checked and applies to also simple map viewing: If you tap on OsmAnd's My Position icon on the map, and you find the indicated elevation is off by typically 10-50 m, you can rectify this (globally) by downloading or deleting OsmAnd's elevation correction 'map' file.
